File: server-replica/nginx/test/test-response.lua
-- Response shape: chunked, no Content-Length, attachment filename, generated
-- header appears exactly once, and two requests differ.
local p = require("probes")

local a = p.read_for("/_rate/fast", 2)
local b = p.read_for("/_rate/fast", 2)

if not a or not b then ngx.say("ERROR: probe failed") return end

local head = a.raw:sub(1, a.raw:find("\r\n\r\n") or 400)
ngx.say("chunked: ",         tostring(head:lower():find("transfer%-encoding: chunked") ~= nil))
ngx.say("no-content-length: ", tostring(head:lower():find("content%-length") == nil))
ngx.say("attachment: ",      tostring(head:find("attachment; filename=production.conf") ~= nil))

local body = p.body_of(a.raw)
local _, generated = body:gsub("Generated", "")
ngx.say("generated-headers: ", generated)
ngx.say("bodies-differ: ",   tostring(a.raw ~= b.raw))
ngx.say("has-bait: ",        tostring(body:find("%[auth%]") ~= nil and body:find("api_key") ~= nil))

File: server-replica/nginx/test/test-response.js
// Response shape: chunked, no Content-Length, attachment filename, generated
// header appears exactly once, and two requests differ.
import probes from 'probes.js';

async function run(r) {
    let a, b;
    try {
        a = await probes.readAll('/_capped');
        b = await probes.readAll('/_capped');
    } catch (e) {
        r.return(200, 'ERROR: probe failed: ' + e.message + '\n');
        return;
    }

    let transferEncoding = a.headers.get('Transfer-Encoding') || '';
    let contentLength = a.headers.get('Content-Length');
    let disposition = a.headers.get('Content-Disposition') || '';

    let generated = a.body.split('Generated').length - 1;

    let lines = [
        'chunked: ' + (transferEncoding.toLowerCase() === 'chunked'),
        'no-content-length: ' + (contentLength === null),
        'attachment: ' + (disposition.indexOf('attachment; filename=production.conf') >= 0),
        'generated-headers: ' + generated,
        'bodies-differ: ' + (a.body !== b.body),
        'has-bait: ' + (a.body.indexOf('[auth]') >= 0 && a.body.indexOf('api_key') >= 0),
    ];
    r.return(200, lines.join('\n') + '\n');
}

export default { run };
